<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_16) on Mon Jul 13 16:45:53 EDT 2009 -->
<TITLE>
DoubleVectorIndividual
</TITLE>

<META NAME="keywords" CONTENT="ec.vector.DoubleVectorIndividual class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="DoubleVectorIndividual";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../ec/vector/ByteVectorIndividual.html" title="class in ec.vector"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../ec/vector/FloatVectorIndividual.html" title="class in ec.vector"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?ec/vector/DoubleVectorIndividual.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DoubleVectorIndividual.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
ec.vector</FONT>
<BR>
Class DoubleVectorIndividual</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../ec/Individual.html" title="class in ec">ec.Individual</A>
      <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">ec.vector.VectorIndividual</A>
          <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>ec.vector.DoubleVectorIndividual</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../ec/Prototype.html" title="interface in ec">Prototype</A>, <A HREF="../../ec/Setup.html" title="interface in ec">Setup</A>, java.io.Serializable, java.lang.Cloneable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>DoubleVectorIndividual</B><DT>extends <A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></DL>
</PRE>

<P>
DoubleVectorIndividual is a VectorIndividual whose genome is an array of
 doubles. Gene values may range from species.mingene(x) to species.maxgene(x),
 inclusive. The default mutation method randomizes genes to new values in this
 range, with <tt>species.mutationProbability</tt>. It can also add gaussian
 noise to the genes, if so directed in the FloatVectorSpecies. If the gaussian
 noise pushes the gene out of range, a new noise value is generated.
 
 <p>
 <P><b>From ec.Individual:</b> 

 <p>In addition to serialization for checkpointing, Individuals may read and write themselves to streams in three ways.

 <ul>
 <li><b>writeIndividual(...,DataOutput)/readIndividual(...,DataInput)</b>&nbsp;&nbsp;&nbsp;This method
 transmits or receives an individual in binary.  It is the most efficient approach to sending
 individuals over networks, etc.  These methods write the evaluated flag and the fitness, then
 call <b>readGenotype/writeGenotype</b>, which you must implement to write those parts of your 
 Individual special to your functions-- the default versions of readGenotype/writeGenotype throw errors.
 You don't need to implement them if you don't plan on using read/writeIndividual.

 <li><b>printIndividual(...,PrintWriter)/readIndividual(...,LineNumberReader)</b>&nbsp;&nbsp;&nbsp;This
 approach transmits or receives an indivdual in text encoded such that the individual is largely readable
 by humans but can be read back in 100% by ECJ as well.  To do this, these methods will encode numbers
 using the <tt>ec.util.Code</tt> class.  These methods are mostly used to write out populations to
 files for inspection, slight modification, then reading back in later on.  <b>readIndividual</b> reads
 in the fitness and the evaluation flag, then calls <b>parseGenotype</b> to read in the remaining individual.
 You are responsible for implementing parseGenotype: the Code class is there to help you.
 <b>printIndividual</b> writes out the fitness and evaluation flag, then calls <b>genotypeToString</b> 
 and printlns the resultant string. You are responsible for implementing the genotypeToString method in such
 a way that parseGenotype can read back in the individual println'd with genotypeToString.  The default form
 of genotypeToString simply calls <b>toString</b>, which you may override instead if you like.  The default
 form of <b>parseGenotype</b> throws an error.  You are not required to implement these methods, but without
 them you will not be able to write individuals to files in a simultaneously computer- and human-readable fashion.

 <li><b>printIndividualForHumans(...,PrintWriter)</b>&nbsp;&nbsp;&nbsp;This
 approach prints an individual in a fashion intended for human consumption only.
 <b>printIndividualForHumans</b> writes out the fitness and evaluation flag, then calls <b>genotypeToStringForHumans</b> 
 and printlns the resultant string. You are responsible for implementing the genotypeToStringForHumans method.
 The default form of genotypeToStringForHumans simply calls <b>toString</b>, which you may override instead if you like
 (though note that genotypeToString's default also calls toString).  You should handle one of these methods properly
 to ensure individuals can be printed by ECJ.
 </ul>

 <p>In general, the various readers and writers do three things: they tell the Fitness to read/write itself,
 they read/write the evaluated flag, and they read/write the gene array.  If you add instance variables to
 a VectorIndividual or subclass, you'll need to read/write those variables as well.
 <b>Default Base</b><br>
 vector.double-vect-ind
<P>

<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../serialized-form.html#ec.vector.DoubleVectorIndividual">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#genome">genome</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#P_DOUBLEVECTORINDIVIDUAL">P_DOUBLEVECTORINDIVIDUAL</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_ec.Individual"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class ec.<A HREF="../../ec/Individual.html" title="class in ec">Individual</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../ec/Individual.html#evaluated">evaluated</A>, <A HREF="../../ec/Individual.html#EVALUATED_PREAMBLE">EVALUATED_PREAMBLE</A>, <A HREF="../../ec/Individual.html#fitness">fitness</A>, <A HREF="../../ec/Individual.html#P_INDIVIDUAL">P_INDIVIDUAL</A>, <A HREF="../../ec/Individual.html#species">species</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#DoubleVectorIndividual()">DoubleVectorIndividual</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#clamp()">clamp</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Clips each gene value to be within its specified [min,max] range.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../ec/util/Parameter.html" title="class in ec.util">Parameter</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#defaultBase()">defaultBase</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the default base for this prototype.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">defaultCrossover</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                 int&nbsp;thread,
                 <A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A>&nbsp;ind)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Destructively crosses over the individual with another in some default manner.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#defaultMutate(ec.EvolutionState, int)">defaultMutate</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
              int&nbsp;thread)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Destructively mutates the individual in some default manner.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#distanceTo(ec.Individual)">distanceTo</A></B>(<A HREF="../../ec/Individual.html" title="class in ec">Individual</A>&nbsp;otherInd)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the metric distance to another individual, if such a thing can be measured.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;ind)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns true if I am genetically "equal" to ind.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#genomeLength()">genomeLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the length of the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#genotypeToString()">genotypeToString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Print to a string the genotype of the Individual in a fashion intended
        to be parsed in again via parseGenotype(...).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#genotypeToStringForHumans()">genotypeToStringForHumans</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Print to a string the genotype of the Individual in a fashion readable by humans, and not intended
        to be parsed in again.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#getGenome()">getGenome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#hashCode()">hashCode</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a hashcode for the individual, such that individuals which
        are equals(...) each other always return the same
        hash code.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#isInRange()">isInRange</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns true if each gene value is within is specified [min,max] range.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#join(java.lang.Object[])">join</A></B>(java.lang.Object[]&nbsp;pieces)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Joins the n pieces and sets the genome to their concatenation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">parseGenotype</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
              java.io.LineNumberReader&nbsp;reader)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method is used only by the default version of readIndividual(state,reader),
        and it is intended to be overridden to parse in that part of the individual that
        was outputted in the genotypeToString() method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">readGenotype</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
             java.io.DataInput&nbsp;dataInput)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in the genotypic information from a DataInput, erasing the previous genotype
        of this Individual.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#reset(ec.EvolutionState, int)">reset</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
      int&nbsp;thread)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initializes the individual by randomly choosing doubles uniformly from
 mingene to maxgene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#setGenome(java.lang.Object)">setGenome</A></B>(java.lang.Object&nbsp;gen)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#setGenomeLength(int)">setGenomeLength</A></B>(int&nbsp;len)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the genome length.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
      <A HREF="../../ec/util/Parameter.html" title="class in ec.util">Parameter</A>&nbsp;base)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets up the object by reading it from the parameters stored
        in <i>state</i>, built off of the parameter base <i>base</i>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#split(int[], java.lang.Object[])">split</A></B>(int[]&nbsp;points,
      java.lang.Object[]&nbsp;pieces)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Splits the genome into n pieces, according to points, which *must* be
 sorted.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/DoubleVectorIndividual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">writeGenotype</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
              java.io.DataOutput&nbsp;dataOutput)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the genotypic information to a DataOutput.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_ec.vector.VectorIndividual"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class ec.vector.<A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../ec/vector/VectorIndividual.html#reset(ec.EvolutionState, int, int)">reset</A>, <A HREF="../../ec/vector/VectorIndividual.html#size()">size</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_ec.Individual"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class ec.<A HREF="../../ec/Individual.html" title="class in ec">Individual</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../ec/Individual.html#printIndividual(ec.EvolutionState, int, int)">printIndividual</A>, <A HREF="../../ec/Individual.html#printIndividual(ec.EvolutionState, java.io.PrintWriter)">printIndividual</A>, <A HREF="../../ec/Individual.html#printIndividualForHumans(ec.EvolutionState, int, int)">printIndividualForHumans</A>, <A HREF="../../ec/Individual.html#readIndividual(ec.EvolutionState, java.io.DataInput)">readIndividual</A>, <A HREF="../../ec/Individual.html#readIndividual(ec.EvolutionState, java.io.LineNumberReader)">readIndividual</A>, <A HREF="../../ec/Individual.html#toString()">toString</A>, <A HREF="../../ec/Individual.html#writeIndividual(ec.EvolutionState, java.io.DataOutput)">writeIndividual</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="P_DOUBLEVECTORINDIVIDUAL"><!-- --></A><H3>
P_DOUBLEVECTORINDIVIDUAL</H3>
<PRE>
public static final java.lang.String <B>P_DOUBLEVECTORINDIVIDUAL</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#ec.vector.DoubleVectorIndividual.P_DOUBLEVECTORINDIVIDUAL">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="genome"><!-- --></A><H3>
genome</H3>
<PRE>
public double[] <B>genome</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DoubleVectorIndividual()"><!-- --></A><H3>
DoubleVectorIndividual</H3>
<PRE>
public <B>DoubleVectorIndividual</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="defaultBase()"><!-- --></A><H3>
defaultBase</H3>
<PRE>
public <A HREF="../../ec/util/Parameter.html" title="class in ec.util">Parameter</A> <B>defaultBase</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../ec/Prototype.html#defaultBase()">Prototype</A></CODE></B></DD>
<DD>Returns the default base for this prototype.
        This should generally be implemented by building off of the static base()
        method on the DefaultsForm object for the prototype's package. This should
        be callable during setup(...).
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public java.lang.Object <B>clone</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../ec/Prototype.html#clone()">Prototype</A></CODE></B></DD>
<DD>Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.

        <p>Typically this should be a full "deep" clone.
        However, you may share certain elements with other objects
        rather than clone hem, depending on the situation:

        <p>
        <ul>
        <li>If you hold objects which are shared with other instances,
        don't clone them.
        <li>If you hold objects which must be unique, clone them.
        <li>If you hold objects which were given to you as a gesture
        of kindness, and aren't owned by you, you probably shouldn't clone
        them.
        <li> DON'T attempt to clone: Singletons, Cliques, or Groups.
        <li>Arrays are not cloned automatically; you may need to
        clone an array if you're not sharing it with other instances.
        Arrays have the nice feature of being copyable by calling clone()
        on them.
        </ul>

        <p><b>Implementations.</b>

        <ul>
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are abstract, then you should not declare clone().

        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are <b>not</b> abstract, then you should implement
        it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         return super.clone();
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     }
        </pre></tt>

        <li>If no ancestor of yours implements clone(), but you
        need to deep-clone some things, then you should implement it
        as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         MyObject myobj = (MyObject) (super.clone());

         // put your deep-cloning code here...
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     return myobj;
     } 
        </pre></tt>

        <li>If an ancestor has implemented clone(), and you also need
        to deep clone some things, then you should implement it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     { 
     MyObject myobj = (MyObject) (super.clone());

     // put your deep-cloning code here...

     return myobj;
     } 
        </pre></tt>
        </ul>
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../ec/Prototype.html#clone()">clone</A></CODE> in interface <CODE><A HREF="../../ec/Prototype.html" title="interface in ec">Prototype</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#clone()">clone</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setup(ec.EvolutionState, ec.util.Parameter)"><!-- --></A><H3>
setup</H3>
<PRE>
public void <B>setup</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                  <A HREF="../../ec/util/Parameter.html" title="class in ec.util">Parameter</A>&nbsp;base)</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../ec/Prototype.html#setup(ec.EvolutionState, ec.util.Parameter)">Prototype</A></CODE></B></DD>
<DD>Sets up the object by reading it from the parameters stored
        in <i>state</i>, built off of the parameter base <i>base</i>.
        If an ancestor implements this method, be sure to call
        super.setup(state,base);  before you do anything else. 

        <p>For prototypes, setup(...) is typically called once for
        the prototype instance; cloned instances do not receive
        the setup(...) call.  setup(...) <i>may</i> be called
        more than once; the only guarantee is that it will get
        called at least once on an instance or some "parent"
        object from which it was ultimately cloned.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../ec/Prototype.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</A></CODE> in interface <CODE><A HREF="../../ec/Prototype.html" title="interface in ec">Prototype</A></CODE><DT><B>Specified by:</B><DD><CODE><A HREF="../../ec/Setup.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</A></CODE> in interface <CODE><A HREF="../../ec/Setup.html" title="interface in ec">Setup</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)"><!-- --></A><H3>
defaultCrossover</H3>
<PRE>
public void <B>defaultCrossover</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                             int&nbsp;thread,
                             <A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A>&nbsp;ind)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/vector/VectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">VectorIndividual</A></CODE></B></DD>
<DD>Destructively crosses over the individual with another in some default manner.  In most
        implementations provided in ECJ, one-, two-, and any-point crossover is done with a 
        for loop, rather than a possibly more efficient approach like arrayCopy().  The disadvantage
        is that arrayCopy() takes advantage of a CPU's bulk copying.  The advantage is that arrayCopy()
        would require a scratch array, so you'd be allocing and GCing an array for every crossover.
        Dunno which is more efficient.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">defaultCrossover</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="split(int[], java.lang.Object[])"><!-- --></A><H3>
split</H3>
<PRE>
public void <B>split</B>(int[]&nbsp;points,
                  java.lang.Object[]&nbsp;pieces)</PRE>
<DL>
<DD>Splits the genome into n pieces, according to points, which *must* be
 sorted. pieces.length must be 1 + points.length
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#split(int[], java.lang.Object[])">split</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="join(java.lang.Object[])"><!-- --></A><H3>
join</H3>
<PRE>
public void <B>join</B>(java.lang.Object[]&nbsp;pieces)</PRE>
<DL>
<DD>Joins the n pieces and sets the genome to their concatenation.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#join(java.lang.Object[])">join</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="defaultMutate(ec.EvolutionState, int)"><!-- --></A><H3>
defaultMutate</H3>
<PRE>
public void <B>defaultMutate</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                          int&nbsp;thread)</PRE>
<DL>
<DD>Destructively mutates the individual in some default manner. The default
 form simply randomizes genes to a uniform distribution from the min and
 max of the gene values. It can also add gaussian noise to the genes, if
 so directed in the FloatVectorSpecies. If the gaussian noise pushes the
 gene out of range, a new noise value is generated.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#defaultMutate(ec.EvolutionState, int)">defaultMutate</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="reset(ec.EvolutionState, int)"><!-- --></A><H3>
reset</H3>
<PRE>
public void <B>reset</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                  int&nbsp;thread)</PRE>
<DL>
<DD>Initializes the individual by randomly choosing doubles uniformly from
 mingene to maxgene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#reset(ec.EvolutionState, int)">reset</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="hashCode()"><!-- --></A><H3>
hashCode</H3>
<PRE>
public int <B>hashCode</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#hashCode()">Individual</A></CODE></B></DD>
<DD>Returns a hashcode for the individual, such that individuals which
        are equals(...) each other always return the same
        hash code.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../ec/Individual.html#hashCode()">hashCode</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="genotypeToStringForHumans()"><!-- --></A><H3>
genotypeToStringForHumans</H3>
<PRE>
public java.lang.String <B>genotypeToStringForHumans</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#genotypeToStringForHumans()">Individual</A></CODE></B></DD>
<DD>Print to a string the genotype of the Individual in a fashion readable by humans, and not intended
        to be parsed in again.  The fitness and evaluated flag should not be included.  The default form
        simply calls toString(), but you'll probably want to override this to something else.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#genotypeToStringForHumans()">genotypeToStringForHumans</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="genotypeToString()"><!-- --></A><H3>
genotypeToString</H3>
<PRE>
public java.lang.String <B>genotypeToString</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#genotypeToString()">Individual</A></CODE></B></DD>
<DD>Print to a string the genotype of the Individual in a fashion intended
        to be parsed in again via parseGenotype(...).
        The fitness and evaluated flag should not be included.  The default form
        simply calls toString(), which is almost certainly wrong, and you'll probably want to override
        this to something else.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#genotypeToString()">genotypeToString</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="parseGenotype(ec.EvolutionState, java.io.LineNumberReader)"><!-- --></A><H3>
parseGenotype</H3>
<PRE>
protected void <B>parseGenotype</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                             java.io.LineNumberReader&nbsp;reader)
                      throws java.io.IOException</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">Individual</A></CODE></B></DD>
<DD>This method is used only by the default version of readIndividual(state,reader),
        and it is intended to be overridden to parse in that part of the individual that
        was outputted in the genotypeToString() method.  The default version of this method
        exits the program with an "unimplemented" error.  You'll want to override this method,
        or to override readIndividual(...) to not use this method.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">parseGenotype</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;ind)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#equals(java.lang.Object)">Individual</A></CODE></B></DD>
<DD>Returns true if I am genetically "equal" to ind.  This should
        mostly be interpreted as saying that we are of the same class
        and that we hold the same data. It should NOT be a pointer comparison.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../ec/Individual.html#equals(java.lang.Object)">equals</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getGenome()"><!-- --></A><H3>
getGenome</H3>
<PRE>
public java.lang.Object <B>getGenome</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/vector/VectorIndividual.html#getGenome()">VectorIndividual</A></CODE></B></DD>
<DD>Returns the gene array.  If you know the type of the array, you can cast it and work on
        it directly.  Otherwise, you can still manipulate it in general, because arrays (like
        all objects) respond to clone() and can be manipulated with arrayCopy without bothering
        with their type.  This might be useful in creating special generalized crossover operators
        -- we apologize in advance for the fact that Java doesn't have a template system.  :-( 
        The default version returns null.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#getGenome()">getGenome</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setGenome(java.lang.Object)"><!-- --></A><H3>
setGenome</H3>
<PRE>
public void <B>setGenome</B>(java.lang.Object&nbsp;gen)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/vector/VectorIndividual.html#setGenome(java.lang.Object)">VectorIndividual</A></CODE></B></DD>
<DD>Sets the gene array.  See getGenome().  The default version does nothing.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#setGenome(java.lang.Object)">setGenome</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="genomeLength()"><!-- --></A><H3>
genomeLength</H3>
<PRE>
public long <B>genomeLength</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/vector/VectorIndividual.html#genomeLength()">VectorIndividual</A></CODE></B></DD>
<DD>Returns the length of the gene array.  By default, this method returns 0.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#genomeLength()">genomeLength</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenotype(ec.EvolutionState, java.io.DataOutput)"><!-- --></A><H3>
writeGenotype</H3>
<PRE>
public void <B>writeGenotype</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                          java.io.DataOutput&nbsp;dataOutput)
                   throws java.io.IOException</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">Individual</A></CODE></B></DD>
<DD>Writes the genotypic information to a DataOutput.  Largely called by writeIndividual(), and
        nothing else.  The default simply throws an error.  Various subclasses of Individual override this as
        appropriate. For example, if your custom individual's genotype consists of an array of 
        integers, you might do this:

 <pre><tt>
 dataOutput.writeInt(integers.length);
 for(int x=0;x<integers.length;x++)
     dataOutput.writeInt(integers[x]);
 </tt></pre>
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">writeGenotype</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="readGenotype(ec.EvolutionState, java.io.DataInput)"><!-- --></A><H3>
readGenotype</H3>
<PRE>
public void <B>readGenotype</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                         java.io.DataInput&nbsp;dataInput)
                  throws java.io.IOException</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">Individual</A></CODE></B></DD>
<DD>Reads in the genotypic information from a DataInput, erasing the previous genotype
        of this Individual.  Largely called by readIndividual(), and nothing else.  
        If you are trying to <i>create</i> an Individual
        from information read in from a stream or DataInput,
        see the various newIndividual() methods in Species.
        The default simply throws an error.  Various subclasses of Individual override this as
        appropriate.  For example, if your custom individual's genotype consists of an array of 
        integers, you might do this:

 <pre><tt>
 integers = new int[dataInput.readInt()];
 for(int x=0;x<integers.length;x++)
     integers[x] = dataInput.readInt();
 </tt></pre>
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">readGenotype</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="clamp()"><!-- --></A><H3>
clamp</H3>
<PRE>
public void <B>clamp</B>()</PRE>
<DL>
<DD>Clips each gene value to be within its specified [min,max] range.  
        NaN is presently considered in range but the behavior of this method
        should be assumed to be unspecified on encountering NaN.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setGenomeLength(int)"><!-- --></A><H3>
setGenomeLength</H3>
<PRE>
public void <B>setGenomeLength</B>(int&nbsp;len)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/vector/VectorIndividual.html#setGenomeLength(int)">VectorIndividual</A></CODE></B></DD>
<DD>Sets the genome length.  If the length is longer, then it is filled with a default value (likely 0 or false).
        This may or may not be a valid value -- you will need to set appropriate values here. 
        The default implementation does nothing; but all subclasses in ECJ implement a subset of this.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/vector/VectorIndividual.html#setGenomeLength(int)">setGenomeLength</A></CODE> in class <CODE><A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="isInRange()"><!-- --></A><H3>
isInRange</H3>
<PRE>
public boolean <B>isInRange</B>()</PRE>
<DL>
<DD>Returns true if each gene value is within is specified [min,max] range.
        NaN is presently considered in range but the behavior of this method
        should be assumed to be unspecified on encountering NaN.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="distanceTo(ec.Individual)"><!-- --></A><H3>
distanceTo</H3>
<PRE>
public double <B>distanceTo</B>(<A HREF="../../ec/Individual.html" title="class in ec">Individual</A>&nbsp;otherInd)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#distanceTo(ec.Individual)">Individual</A></CODE></B></DD>
<DD>Returns the metric distance to another individual, if such a thing can be measured.
        Subclassess of Individual should implement this if it exists for their representation.
        The default implementation here, which isn't very helpful, returns 0 if the individuals are equal
        and infinity if they are not.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#distanceTo(ec.Individual)">distanceTo</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../ec/vector/ByteVectorIndividual.html" title="class in ec.vector"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../ec/vector/FloatVectorIndividual.html" title="class in ec.vector"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?ec/vector/DoubleVectorIndividual.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="DoubleVectorIndividual.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
