<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_16) on Mon Jul 13 16:45:53 EDT 2009 -->
<TITLE>
VectorIndividual
</TITLE>

<META NAME="keywords" CONTENT="ec.vector.VectorIndividual class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="VectorIndividual";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../ec/vector/VectorGene.html" title="class in ec.vector"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../ec/vector/VectorSpecies.html" title="class in ec.vector"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?ec/vector/VectorIndividual.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="VectorIndividual.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#fields_inherited_from_class_ec.Individual">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
ec.vector</FONT>
<BR>
Class VectorIndividual</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../ec/Individual.html" title="class in ec">ec.Individual</A>
      <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>ec.vector.VectorIndividual</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../ec/Prototype.html" title="interface in ec">Prototype</A>, <A HREF="../../ec/Setup.html" title="interface in ec">Setup</A>, java.io.Serializable, java.lang.Cloneable</DD>
</DL>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../../ec/vector/BitVectorIndividual.html" title="class in ec.vector">BitVectorIndividual</A>, <A HREF="../../ec/vector/ByteVectorIndividual.html" title="class in ec.vector">ByteVectorIndividual</A>, <A HREF="../../ec/vector/DoubleVectorIndividual.html" title="class in ec.vector">DoubleVectorIndividual</A>, <A HREF="../../ec/vector/FloatVectorIndividual.html" title="class in ec.vector">FloatVectorIndividual</A>, <A HREF="../../ec/vector/GeneVectorIndividual.html" title="class in ec.vector">GeneVectorIndividual</A>, <A HREF="../../ec/vector/IntegerVectorIndividual.html" title="class in ec.vector">IntegerVectorIndividual</A>, <A HREF="../../ec/vector/LongVectorIndividual.html" title="class in ec.vector">LongVectorIndividual</A>, <A HREF="../../ec/vector/ShortVectorIndividual.html" title="class in ec.vector">ShortVectorIndividual</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public abstract class <B>VectorIndividual</B><DT>extends <A HREF="../../ec/Individual.html" title="class in ec">Individual</A></DL>
</PRE>

<P>
VectorIndividual is the abstract superclass of simple individual representations
 which consist of vectors of values (booleans, integers, floating-point, etc.)

 <p>This class contains two methods, defaultCrossover and defaultMutate, which can
 be overridden if all you need is a simple crossover and a simple mutate mechanism.
 the VectorCrossoverPipeline and VectorMutationPipeline classes use these methods to do their
 handiwork.  For more sophisticated crossover and mutation, you'll need to write
 a custom breeding pipeline.

 <p>The <i>kind</i> of default crossover and mutation, and associated information,
 is stored in the VectorIndividual's VectorSpecies object, which is obtained through
 the <tt>species</tt> variable.  For example, 
 VectorIndividual assumes three common types of crossover as defined in VectorSpecies
 which you should implement in your defaultCrossover method: one-point, 
 two-point, and any-point (otherwise known as "uniform") crossover.

 <p>VectorIndividual is typically used for fixed-length vector representations;
 however, it can also be used with variable-length representations.  Two methods have
 been provided in all subclasses of VectorIndividual to help you there: split and
 join, which you can use to break up and reconnect VectorIndividuals in a variety
 of ways.  Note that you may want to override the reset() method to create individuals
 with different initial lengths.

 <p>VectorIndividuals must belong to the species VectorSpecies (or some subclass of it).


 <P><b>From ec.Individual:</b>

 <p>In addition to serialization for checkpointing, Individuals may read and write themselves to streams in three ways.

 <ul>
 <li><b>writeIndividual(...,DataOutput)/readIndividual(...,DataInput)</b>&nbsp;&nbsp;&nbsp;This method
 transmits or receives an individual in binary.  It is the most efficient approach to sending
 individuals over networks, etc.  These methods write the evaluated flag and the fitness, then
 call <b>readGenotype/writeGenotype</b>, which you must implement to write those parts of your 
 Individual special to your functions-- the default versions of readGenotype/writeGenotype throw errors.
 You don't need to implement them if you don't plan on using read/writeIndividual.

 <li><b>printIndividual(...,PrintWriter)/readIndividual(...,LineNumberReader)</b>&nbsp;&nbsp;&nbsp;This
 approach transmits or receives an indivdual in text encoded such that the individual is largely readable
 by humans but can be read back in 100% by ECJ as well.  To do this, these methods will encode numbers
 using the <tt>ec.util.Code</tt> class.  These methods are mostly used to write out populations to
 files for inspection, slight modification, then reading back in later on.  <b>readIndividual</b> reads
 in the fitness and the evaluation flag, then calls <b>parseGenotype</b> to read in the remaining individual.
 You are responsible for implementing parseGenotype: the Code class is there to help you.
 <b>printIndividual</b> writes out the fitness and evaluation flag, then calls <b>genotypeToString</b> 
 and printlns the resultant string. You are responsible for implementing the genotypeToString method in such
 a way that parseGenotype can read back in the individual println'd with genotypeToString.  The default form
 of genotypeToString simply calls <b>toString</b>, which you may override instead if you like.  The default
 form of <b>parseGenotype</b> throws an error.  You are not required to implement these methods, but without
 them you will not be able to write individuals to files in a simultaneously computer- and human-readable fashion.

 <li><b>printIndividualForHumans(...,PrintWriter)</b>&nbsp;&nbsp;&nbsp;This
 approach prints an individual in a fashion intended for human consumption only.
 <b>printIndividualForHumans</b> writes out the fitness and evaluation flag, then calls <b>genotypeToStringForHumans</b> 
 and printlns the resultant string. You are responsible for implementing the genotypeToStringForHumans method.
 The default form of genotypeToStringForHumans simply calls <b>toString</b>, which you may override instead if you like
 (though note that genotypeToString's default also calls toString).  You should handle one of these methods properly
 to ensure individuals can be printed by ECJ.
 </ul>

 <p>In general, the various readers and writers do three things: they tell the Fitness to read/write itself,
 they read/write the evaluated flag, and they read/write the gene array.  If you add instance variables to
 a VectorIndividual or subclass, you'll need to read/write those variables as well.
<P>

<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../serialized-form.html#ec.vector.VectorIndividual">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_ec.Individual"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class ec.<A HREF="../../ec/Individual.html" title="class in ec">Individual</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../ec/Individual.html#evaluated">evaluated</A>, <A HREF="../../ec/Individual.html#EVALUATED_PREAMBLE">EVALUATED_PREAMBLE</A>, <A HREF="../../ec/Individual.html#fitness">fitness</A>, <A HREF="../../ec/Individual.html#P_INDIVIDUAL">P_INDIVIDUAL</A>, <A HREF="../../ec/Individual.html#species">species</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#VectorIndividual()">VectorIndividual</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">defaultCrossover</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                 int&nbsp;thread,
                 <A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A>&nbsp;ind)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Destructively crosses over the individual with another in some default manner.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#defaultMutate(ec.EvolutionState, int)">defaultMutate</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
              int&nbsp;thread)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Destructively mutates the individual in some default manner.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#genomeLength()">genomeLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the length of the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#getGenome()">getGenome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#join(java.lang.Object[])">join</A></B>(java.lang.Object[]&nbsp;pieces)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Joins the n pieces and sets the genome to their concatenation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>abstract &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#reset(ec.EvolutionState, int)">reset</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
      int&nbsp;thread)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initializes the individual.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#reset(ec.EvolutionState, int, int)">reset</A></B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
      int&nbsp;thread,
      int&nbsp;newSize)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initializes the individual to a new size.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#setGenome(java.lang.Object)">setGenome</A></B>(java.lang.Object&nbsp;gen)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#setGenomeLength(int)">setGenomeLength</A></B>(int&nbsp;len)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the genome length.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#size()">size</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the "size" of the individual.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ec/vector/VectorIndividual.html#split(int[], java.lang.Object[])">split</A></B>(int[]&nbsp;points,
      java.lang.Object[]&nbsp;pieces)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Splits the genome into n pieces, according to points, which *must* be sorted.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_ec.Individual"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class ec.<A HREF="../../ec/Individual.html" title="class in ec">Individual</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../ec/Individual.html#clone()">clone</A>, <A HREF="../../ec/Individual.html#distanceTo(ec.Individual)">distanceTo</A>, <A HREF="../../ec/Individual.html#equals(java.lang.Object)">equals</A>, <A HREF="../../ec/Individual.html#genotypeToString()">genotypeToString</A>, <A HREF="../../ec/Individual.html#genotypeToStringForHumans()">genotypeToStringForHumans</A>, <A HREF="../../ec/Individual.html#hashCode()">hashCode</A>, <A HREF="../../ec/Individual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">parseGenotype</A>, <A HREF="../../ec/Individual.html#printIndividual(ec.EvolutionState, int, int)">printIndividual</A>, <A HREF="../../ec/Individual.html#printIndividual(ec.EvolutionState, java.io.PrintWriter)">printIndividual</A>, <A HREF="../../ec/Individual.html#printIndividualForHumans(ec.EvolutionState, int, int)">printIndividualForHumans</A>, <A HREF="../../ec/Individual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">readGenotype</A>, <A HREF="../../ec/Individual.html#readIndividual(ec.EvolutionState, java.io.DataInput)">readIndividual</A>, <A HREF="../../ec/Individual.html#readIndividual(ec.EvolutionState, java.io.LineNumberReader)">readIndividual</A>, <A HREF="../../ec/Individual.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</A>, <A HREF="../../ec/Individual.html#toString()">toString</A>, <A HREF="../../ec/Individual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">writeGenotype</A>, <A HREF="../../ec/Individual.html#writeIndividual(ec.EvolutionState, java.io.DataOutput)">writeIndividual</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_ec.Prototype"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface ec.<A HREF="../../ec/Prototype.html" title="interface in ec">Prototype</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../ec/Prototype.html#defaultBase()">defaultBase</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="VectorIndividual()"><!-- --></A><H3>
VectorIndividual</H3>
<PRE>
public <B>VectorIndividual</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)"><!-- --></A><H3>
defaultCrossover</H3>
<PRE>
public void <B>defaultCrossover</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                             int&nbsp;thread,
                             <A HREF="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</A>&nbsp;ind)</PRE>
<DL>
<DD>Destructively crosses over the individual with another in some default manner.  In most
        implementations provided in ECJ, one-, two-, and any-point crossover is done with a 
        for loop, rather than a possibly more efficient approach like arrayCopy().  The disadvantage
        is that arrayCopy() takes advantage of a CPU's bulk copying.  The advantage is that arrayCopy()
        would require a scratch array, so you'd be allocing and GCing an array for every crossover.
        Dunno which is more efficient.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="defaultMutate(ec.EvolutionState, int)"><!-- --></A><H3>
defaultMutate</H3>
<PRE>
public void <B>defaultMutate</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                          int&nbsp;thread)</PRE>
<DL>
<DD>Destructively mutates the individual in some default manner.  The default version calls reset()
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="reset(ec.EvolutionState, int)"><!-- --></A><H3>
reset</H3>
<PRE>
public abstract void <B>reset</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                           int&nbsp;thread)</PRE>
<DL>
<DD>Initializes the individual.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getGenome()"><!-- --></A><H3>
getGenome</H3>
<PRE>
public java.lang.Object <B>getGenome</B>()</PRE>
<DL>
<DD>Returns the gene array.  If you know the type of the array, you can cast it and work on
        it directly.  Otherwise, you can still manipulate it in general, because arrays (like
        all objects) respond to clone() and can be manipulated with arrayCopy without bothering
        with their type.  This might be useful in creating special generalized crossover operators
        -- we apologize in advance for the fact that Java doesn't have a template system.  :-( 
        The default version returns null.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setGenome(java.lang.Object)"><!-- --></A><H3>
setGenome</H3>
<PRE>
public void <B>setGenome</B>(java.lang.Object&nbsp;gen)</PRE>
<DL>
<DD>Sets the gene array.  See getGenome().  The default version does nothing.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="genomeLength()"><!-- --></A><H3>
genomeLength</H3>
<PRE>
public long <B>genomeLength</B>()</PRE>
<DL>
<DD>Returns the length of the gene array.  By default, this method returns 0.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="reset(ec.EvolutionState, int, int)"><!-- --></A><H3>
reset</H3>
<PRE>
public void <B>reset</B>(<A HREF="../../ec/EvolutionState.html" title="class in ec">EvolutionState</A>&nbsp;state,
                  int&nbsp;thread,
                  int&nbsp;newSize)</PRE>
<DL>
<DD>Initializes the individual to a new size.  Only use this if you need to initialize variable-length individuals.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setGenomeLength(int)"><!-- --></A><H3>
setGenomeLength</H3>
<PRE>
public void <B>setGenomeLength</B>(int&nbsp;len)</PRE>
<DL>
<DD>Sets the genome length.  If the length is longer, then it is filled with a default value (likely 0 or false).
        This may or may not be a valid value -- you will need to set appropriate values here. 
        The default implementation does nothing; but all subclasses in ECJ implement a subset of this.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="split(int[], java.lang.Object[])"><!-- --></A><H3>
split</H3>
<PRE>
public void <B>split</B>(int[]&nbsp;points,
                  java.lang.Object[]&nbsp;pieces)</PRE>
<DL>
<DD>Splits the genome into n pieces, according to points, which *must* be sorted. 
        pieces.length must be 1 + points.length.  The default form does nothing -- be careful
        not to use this method if it's not implemented!  It should be trivial to implement it
        for your genome -- just like at the other implementations.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="join(java.lang.Object[])"><!-- --></A><H3>
join</H3>
<PRE>
public void <B>join</B>(java.lang.Object[]&nbsp;pieces)</PRE>
<DL>
<DD>Joins the n pieces and sets the genome to their concatenation.  The default form does nothing. 
        It should be trivial to implement it
        for your genome -- just like at the other implementations.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="size()"><!-- --></A><H3>
size</H3>
<PRE>
public long <B>size</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../ec/Individual.html#size()">Individual</A></CODE></B></DD>
<DD>Returns the "size" of the individual.  This is used for things like
        parsimony pressure.  The default form of this method returns 0 --
        if you care about parsimony pressure, you'll need to override the
        default to provide a more descriptive measure of size.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../ec/Individual.html#size()">size</A></CODE> in class <CODE><A HREF="../../ec/Individual.html" title="class in ec">Individual</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../ec/vector/VectorGene.html" title="class in ec.vector"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../ec/vector/VectorSpecies.html" title="class in ec.vector"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?ec/vector/VectorIndividual.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="VectorIndividual.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#fields_inherited_from_class_ec.Individual">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
