<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_51) on Fri Jun 19 12:54:59 EDT 2015 -->
<title>VectorSpecies</title>
<meta name="date" content="2015-06-19">
<link rel="stylesheet" type="text/css" href="../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="VectorSpecies";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-all.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector"><span class="strong">Prev Class</span></a></li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../index.html?ec/vector/VectorSpecies.html" target="_top">Frames</a></li>
<li><a href="VectorSpecies.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field_summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field_detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">ec.vector</div>
<h2 title="Class VectorSpecies" class="title">Class VectorSpecies</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li><a href="../../ec/Species.html" title="class in ec">ec.Species</a></li>
<li>
<ul class="inheritance">
<li>ec.vector.VectorSpecies</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="../../ec/Prototype.html" title="interface in ec">Prototype</a>, <a href="../../ec/Setup.html" title="interface in ec">Setup</a>, java.io.Serializable, java.lang.Cloneable</dd>
</dl>
<dl>
<dt>Direct Known Subclasses:</dt>
<dd><a href="../../ec/vector/BitVectorSpecies.html" title="class in ec.vector">BitVectorSpecies</a>, <a href="../../ec/vector/FloatVectorSpecies.html" title="class in ec.vector">FloatVectorSpecies</a>, <a href="../../ec/vector/GeneVectorSpecies.html" title="class in ec.vector">GeneVectorSpecies</a>, <a href="../../ec/vector/IntegerVectorSpecies.html" title="class in ec.vector">IntegerVectorSpecies</a></dd>
</dl>
<hr>
<br>
<pre>public class <span class="strong">VectorSpecies</span>
extends <a href="../../ec/Species.html" title="class in ec">Species</a></pre>
<div class="block">VectorSpecies is a species which can create VectorIndividuals.  Different
 VectorSpecies are used for different kinds of VectorIndividuals: a plain
 VectorSpecies is probably only applicable for BitVectorIndividuals.
 
 <p>VectorSpecies supports the following recombination methods:</p>
 <ul>
 <li><b>One-point crossover</b>.</li>
 <li><b>Two-point crossover</b>.</li>
 <li><b>Uniform crossover</b> - inaccurately called "any-point".</li>
 <li><b>Line recombination</b> - children are random points on a line between
      the two parents.</li>
 <li><b>Intermediate recombination</b> - the value of each component of the
      vector is between the values of that component of the parent vectors.
      </li>
 </ul>
 
 <P>Note that BitVectorIndividuals (which use VectorSpecies) and GeneVectorIndividuals
 (which use GeneVectorSpecies, a subclass of VectorSpecies) do not support
 Line or Intermediate Recombination.

 <p>Also note that for LongVectorIndividuals, there are certain values that will
 never be created by line and intermediate recombination, because the
 recombination is calculated using doubles and then rounded to the nearest
 long. For large enough values (but still smaller than the maximum long), the
 difference between one double and the next is greater than one.</p>

 <p>VectorSpecies has three wasy to determine the initial size of the individual:</p>
 <ul>
 <li><b>A fixed size</b>.</li>
 <li><b>Geometric distribution</b>.</li>
 <li><b>Uniform distribution</b></li>
 </ul>

 <p>If the algorithm used is the geometric distribution, the VectorSpecies starts at a
 minimum size and continues flipping a coin with a certain "resize probability",
 increasing the size each time, until the coin comes up tails (fails).  The chunk size
 must be 1 in this case.

 <p> If the algorithm used is the uniform distribution, the VectorSpecies picks a random
 size between a provided minimum and maximum size, inclusive.  The chunk size
 must be 1 in this case.

 <p>If the size is fixed, then you can also provide a "chunk size" which constrains the
 locations in which crossover can be performed (only along chunk boundaries).  The genome
 size must be a multiple of the chunk size in this case.

 <p>VectorSpecies also contains a number of parameters guiding how the individual
 crosses over and mutates.

 <p><b>Per-Gene and Per-Segment Specification.</b>  VectorSpecies and its subclasses
 specify a lot of parameters, notably mutation and initialization parameters, in one
 of three ways.  We will use the <b><tt>mutation-probability</tt></b>
 parameter as an example.

 <ol>
 <li> Globally for all genes in the genome.
      This is done by specifying:
      <p><i>base</i>.<tt>mutation-probability</tt>
      <br><i>base</i>.<tt>max-gene</tt>
      <p><i>Note:</i> you <b>must</b> provide these values even if you don't use them,
      as they're used as defaults by #2 and #3 below.
<p>
 <li> You may provide parameters for genes in segments (regions) along
      the genome.  The idea is to allow you to specify large chunks of genes
      all having the same parameter features.  
      To do this you must first specify how many segments there are:
      <p><i>base</i>.<tt>num-segments</tt>
      <p>The segments then may be defined by either start or end indices of genes. 
      This is controlled by specifying the value of:
      <p><i>base</i>.<tt>segment-type</tt>
      <p>...which can assume the value of start or end, with start being the default.
      The indices are defined using Java array style, i.e. the first gene has the index of 0, 
      and the last gene has the index of genome-size - 1.
      <p>Using this method, each segment is specified by<i>j</i>...
      <p><i>base</i>.<tt>segment.</tt><i>j</i><tt>.start</tt>
      <br><i>base</i>.<tt>segment.</tt><i>j</i><tt>.mutation-probability</tt>
      if segment-type value was chosen as start or by:
      <p><i>base</i>.<tt>segment.</tt><i>j</i><tt>.end</tt>
      <br><i>base</i>.<tt>segment.</tt><i>j</i><tt>.mutation-probability</tt>
      if segment-type value is equal to end.
<p>
 <li> You may parameters for each separate gene.  
      This is done by specifying (for each gene location <i>i</i> you wish to specify)
      <p><i>base</i>.<tt>mutation-probability</tt>.<i>i</i>
 </ol>
 
 <p>Any settings for #3 override #2, and both override #1. 

 <p>The only parameter which can be specified this way in VectorSpecies is at present
 <tt>mutation-probability</tt>.  However a number of parameters are specified this way
 in subclasses. 

 <p><b>Parameters</b><br>
 <table>
 <tr><td valign=top><i>base</i>.<tt>genome-size</tt><br>
 <font size=-1>int &gt;= 1 or one of: geometric, uniform</font></td>
 <td valign=top>(size of the genome, or if 'geometric' or 'uniform', the algorithm used to size the initial genome)</td></tr>

 <tr><td valign=top><i>base</i>.<tt>chunk-size</tt><br>
 <font size=-1>1 &lt;= int &lt;= genome-size (default=1)</font></td>
 <td valign=top>(the chunk size for crossover (crossover will only occur on chunk boundaries))</td></tr>

 <tr><td valign=top><i>base</i>.<tt>geometric-prob</tt><br>
 <font size=-1>0.0 &lt;= double &lt; 1.0</font></td>
 <td valign=top>(the coin-flip probability for increasing the initial size using the geometric distribution)</td></tr>

 <tr><td valign=top><i>base</i>.<tt>min-initial-size</tt><br>
 <font size=-1>int &gt;= 0</font></td>
 <td valign=top>(the minimum initial size of the genome)</td></tr>

 <tr><td valign=top><i>base</i>.<tt>max-initial-size</tt><br>
 <font size=-1>int &gt;= min-initial-size</font></td>
 <td valign=top>(the maximum initial size of the genome)</td></tr>

 <tr><td valign=top><i>base</i>.<tt>crossover-type</tt><br>
 <font size=-1>string, one of: one, two, any</font></td>
 <td valign=top>(default crossover type (one-point, one-point-nonempty, two-point, two-point-nonempty, any-point (uniform), line, or intermediate)</td></tr>

 <tr><td valign=top><i>base</i>.<tt>crossover-prob</tt><br>
 <font size=-1>0.0 &gt;= double &gt;= 1.0 </font></td>
 <td valign=top>(probability that a gene will get crossed over during any-point (uniform) or simulated binary crossover)</td></tr>

 <tr><td valign=top><i>base</i>.<tt>line-extension</tt><br>
 <font size=-1>double &gt;= 0.0 </font></td>
 <td valign=top>(for line and intermediate recombination, how far along the line or outside of the hypercube children can be. If this value is zero, all children must be within the hypercube.)


 <tr><td>&nbsp;
 <tr><td valign=top style="white-space: nowrap"><i>base</i>.<tt>mutation-prob</tt>&nbsp;&nbsp;&nbsp;<i>or</i><br>
 <tr><td valign=top style="white-space: nowrap"><i>base</i>.<tt>segment</tt>.<i>segment-number</i>.<tt>mutation-prob</tt>&nbsp;&nbsp;&nbsp;<i>or</i><br>
 <tr><td valign=top style="white-space: nowrap"><i>base</i>.<tt>mutation-prob</tt>.<i>gene-number</i><br>
 <font size=-1>0.0 &lt;= double &lt;= 1.0 </font></td>
 <td valign=top>(probability that a gene will get mutated over default mutation)</td></tr>

 </table>

 <p><b>Default Base</b><br>
 vector.species</div>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../serialized-form.html#ec.vector.VectorSpecies">Serialized Form</a></dd></dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field_summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_ANY_POINT">C_ANY_POINT</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_GEOMETRIC">C_GEOMETRIC</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_INTERMED_RECOMB">C_INTERMED_RECOMB</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_LINE_RECOMB">C_LINE_RECOMB</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_NONE">C_NONE</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_ONE_POINT">C_ONE_POINT</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_ONE_POINT_NO_NOP">C_ONE_POINT_NO_NOP</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_SIMULATED_BINARY">C_SIMULATED_BINARY</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_TWO_POINT">C_TWO_POINT</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_TWO_POINT_NO_NOP">C_TWO_POINT_NO_NOP</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#C_UNIFORM">C_UNIFORM</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#chunksize">chunksize</a></strong></code>
<div class="block">How big of chunks should we define for crossover?</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#crossoverDistributionIndex">crossoverDistributionIndex</a></strong></code>
<div class="block">What should the SBX distribution index be?</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#crossoverProbability">crossoverProbability</a></strong></code>
<div class="block">Probability that a gene will cross over -- ONLY used in V_ANY_POINT crossover</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#crossoverType">crossoverType</a></strong></code>
<div class="block">What kind of crossover do we have?</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected int[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#duplicateRetries">duplicateRetries</a></strong></code>
<div class="block">How often do we retry until we get a non-duplicate gene?</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#dynamicInitialSize">dynamicInitialSize</a></strong></code>
<div class="block">Was the initial size determined dynamically?</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#genomeIncreaseProbability">genomeIncreaseProbability</a></strong></code>
<div class="block">With what probability would our genome be at least 1 larger than it is now during initialization?</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#genomeResizeAlgorithm">genomeResizeAlgorithm</a></strong></code>
<div class="block">How should we reset the genome?</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#genomeSize">genomeSize</a></strong></code>
<div class="block">How big of a genome should we create on initialization?</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#lineDistance">lineDistance</a></strong></code>
<div class="block">How far along the long a child can be located for line or intermediate recombination</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#maxInitialSize">maxInitialSize</a></strong></code>
<div class="block">What's the largest legal genome?</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#minInitialSize">minInitialSize</a></strong></code>
<div class="block">What's the smallest legal genome?</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected double[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#mutationProbability">mutationProbability</a></strong></code>
<div class="block">Probability that a gene will mutate, per gene.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_CHUNKSIZE">P_CHUNKSIZE</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_CROSSOVER_DISTRIBUTION_INDEX">P_CROSSOVER_DISTRIBUTION_INDEX</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_CROSSOVERPROB">P_CROSSOVERPROB</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_CROSSOVERTYPE">P_CROSSOVERTYPE</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_DUPLICATE_RETRIES">P_DUPLICATE_RETRIES</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_GENOMESIZE">P_GENOMESIZE</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_GEOMETRIC_PROBABILITY">P_GEOMETRIC_PROBABILITY</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_LINEDISTANCE">P_LINEDISTANCE</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_MUTATIONPROB">P_MUTATIONPROB</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_NUM_SEGMENTS">P_NUM_SEGMENTS</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_SEGMENT">P_SEGMENT</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_SEGMENT_END">P_SEGMENT_END</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_SEGMENT_START">P_SEGMENT_START</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_SEGMENT_TYPE">P_SEGMENT_TYPE</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_UNIFORM_MAX">P_UNIFORM_MAX</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_UNIFORM_MIN">P_UNIFORM_MIN</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#P_VECTORSPECIES">P_VECTORSPECIES</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_ANY_POINT">V_ANY_POINT</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_GEOMETRIC">V_GEOMETRIC</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_INTERMED_RECOMB">V_INTERMED_RECOMB</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_LINE_RECOMB">V_LINE_RECOMB</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_ONE_POINT">V_ONE_POINT</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_ONE_POINT_NO_NOP">V_ONE_POINT_NO_NOP</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_SIMULATED_BINARY">V_SIMULATED_BINARY</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_TWO_POINT">V_TWO_POINT</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_TWO_POINT_NO_NOP">V_TWO_POINT_NO_NOP</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#V_UNIFORM">V_UNIFORM</a></strong></code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="fields_inherited_from_class_ec.Species">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;ec.<a href="../../ec/Species.html" title="class in ec">Species</a></h3>
<code><a href="../../ec/Species.html#f_prototype">f_prototype</a>, <a href="../../ec/Species.html#i_prototype">i_prototype</a>, <a href="../../ec/Species.html#P_FITNESS">P_FITNESS</a>, <a href="../../ec/Species.html#P_INDIVIDUAL">P_INDIVIDUAL</a>, <a href="../../ec/Species.html#P_PIPE">P_PIPE</a>, <a href="../../ec/Species.html#pipe_prototype">pipe_prototype</a></code></li>
</ul>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../ec/vector/VectorSpecies.html#VectorSpecies()">VectorSpecies</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#contains(boolean[], boolean)">contains</a></strong>(boolean[]&nbsp;array,
        boolean&nbsp;val)</code>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#contains(double[], double)">contains</a></strong>(double[]&nbsp;array,
        double&nbsp;val)</code>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#contains(int[], int)">contains</a></strong>(int[]&nbsp;array,
        int&nbsp;val)</code>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#contains(long[], long)">contains</a></strong>(long[]&nbsp;array,
        long&nbsp;val)</code>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a></code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#defaultBase()">defaultBase</a></strong>()</code>
<div class="block">Returns the default base for this prototype.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#duplicateRetries(int)">duplicateRetries</a></strong>(int&nbsp;gene)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected boolean[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#fill(boolean[], boolean)">fill</a></strong>(boolean[]&nbsp;array,
    boolean&nbsp;val)</code>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected double[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#fill(double[], double)">fill</a></strong>(double[]&nbsp;array,
    double&nbsp;val)</code>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected int[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#fill(int[], int)">fill</a></strong>(int[]&nbsp;array,
    int&nbsp;val)</code>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected long[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#fill(long[], long)">fill</a></strong>(long[]&nbsp;array,
    long&nbsp;val)</code>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#initializeGenomeSegmentsByEndIndices(ec.EvolutionState, ec.util.Parameter, ec.util.Parameter, int)">initializeGenomeSegmentsByEndIndices</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                                    <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base,
                                    <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;def,
                                    int&nbsp;numSegments)</code>
<div class="block">Looks up genome segments using end indices.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#initializeGenomeSegmentsByStartIndices(ec.EvolutionState, ec.util.Parameter, ec.util.Parameter, int)">initializeGenomeSegmentsByStartIndices</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                                      <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base,
                                      <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;def,
                                      int&nbsp;numSegments)</code>
<div class="block">Looks up genome segments using start indices.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#loadParametersForGene(ec.EvolutionState, int, ec.util.Parameter, ec.util.Parameter, java.lang.String)">loadParametersForGene</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                     int&nbsp;index,
                     <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base,
                     <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;def,
                     java.lang.String&nbsp;postfix)</code>
<div class="block">Called when VectorSpecies is setting up per-gene and per-segment parameters.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#mutationProbability(int)">mutationProbability</a></strong>(int&nbsp;gene)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#newIndividual(ec.EvolutionState, int)">newIndividual</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
             int&nbsp;thread)</code>
<div class="block">Provides a brand-new individual to fill in a population.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
     <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</code>
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/VectorSpecies.html#setupGenome(ec.EvolutionState, ec.util.Parameter)">setupGenome</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
           <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_ec.Species">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.<a href="../../ec/Species.html" title="class in ec">Species</a></h3>
<code><a href="../../ec/Species.html#clone()">clone</a>, <a href="../../ec/Species.html#newIndividual(ec.EvolutionState, java.io.DataInput)">newIndividual</a>, <a href="../../ec/Species.html#newIndividual(ec.EvolutionState, java.io.LineNumberReader)">newIndividual</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field_detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="P_VECTORSPECIES">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_VECTORSPECIES</h4>
<pre>public static final&nbsp;java.lang.String P_VECTORSPECIES</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_VECTORSPECIES">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_CROSSOVERTYPE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CROSSOVERTYPE</h4>
<pre>public static final&nbsp;java.lang.String P_CROSSOVERTYPE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_CROSSOVERTYPE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_CHUNKSIZE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CHUNKSIZE</h4>
<pre>public static final&nbsp;java.lang.String P_CHUNKSIZE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_CHUNKSIZE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_ONE_POINT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_ONE_POINT</h4>
<pre>public static final&nbsp;java.lang.String V_ONE_POINT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_ONE_POINT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_ONE_POINT_NO_NOP">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_ONE_POINT_NO_NOP</h4>
<pre>public static final&nbsp;java.lang.String V_ONE_POINT_NO_NOP</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_ONE_POINT_NO_NOP">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_TWO_POINT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_TWO_POINT</h4>
<pre>public static final&nbsp;java.lang.String V_TWO_POINT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_TWO_POINT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_TWO_POINT_NO_NOP">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_TWO_POINT_NO_NOP</h4>
<pre>public static final&nbsp;java.lang.String V_TWO_POINT_NO_NOP</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_TWO_POINT_NO_NOP">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_ANY_POINT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_ANY_POINT</h4>
<pre>public static final&nbsp;java.lang.String V_ANY_POINT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_ANY_POINT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_LINE_RECOMB">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_LINE_RECOMB</h4>
<pre>public static final&nbsp;java.lang.String V_LINE_RECOMB</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_LINE_RECOMB">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_INTERMED_RECOMB">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_INTERMED_RECOMB</h4>
<pre>public static final&nbsp;java.lang.String V_INTERMED_RECOMB</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_INTERMED_RECOMB">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_SIMULATED_BINARY">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_SIMULATED_BINARY</h4>
<pre>public static final&nbsp;java.lang.String V_SIMULATED_BINARY</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_SIMULATED_BINARY">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_CROSSOVER_DISTRIBUTION_INDEX">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CROSSOVER_DISTRIBUTION_INDEX</h4>
<pre>public static final&nbsp;java.lang.String P_CROSSOVER_DISTRIBUTION_INDEX</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_CROSSOVER_DISTRIBUTION_INDEX">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_MUTATIONPROB">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_MUTATIONPROB</h4>
<pre>public static final&nbsp;java.lang.String P_MUTATIONPROB</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_MUTATIONPROB">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_CROSSOVERPROB">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_CROSSOVERPROB</h4>
<pre>public static final&nbsp;java.lang.String P_CROSSOVERPROB</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_CROSSOVERPROB">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_GENOMESIZE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_GENOMESIZE</h4>
<pre>public static final&nbsp;java.lang.String P_GENOMESIZE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_GENOMESIZE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_LINEDISTANCE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_LINEDISTANCE</h4>
<pre>public static final&nbsp;java.lang.String P_LINEDISTANCE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_LINEDISTANCE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_GEOMETRIC">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_GEOMETRIC</h4>
<pre>public static final&nbsp;java.lang.String V_GEOMETRIC</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_GEOMETRIC">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_GEOMETRIC_PROBABILITY">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_GEOMETRIC_PROBABILITY</h4>
<pre>public static final&nbsp;java.lang.String P_GEOMETRIC_PROBABILITY</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_GEOMETRIC_PROBABILITY">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="V_UNIFORM">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>V_UNIFORM</h4>
<pre>public static final&nbsp;java.lang.String V_UNIFORM</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.V_UNIFORM">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_UNIFORM_MIN">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_UNIFORM_MIN</h4>
<pre>public static final&nbsp;java.lang.String P_UNIFORM_MIN</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_UNIFORM_MIN">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_UNIFORM_MAX">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_UNIFORM_MAX</h4>
<pre>public static final&nbsp;java.lang.String P_UNIFORM_MAX</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_UNIFORM_MAX">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_NUM_SEGMENTS">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_NUM_SEGMENTS</h4>
<pre>public static final&nbsp;java.lang.String P_NUM_SEGMENTS</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_NUM_SEGMENTS">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_SEGMENT_TYPE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_SEGMENT_TYPE</h4>
<pre>public static final&nbsp;java.lang.String P_SEGMENT_TYPE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_SEGMENT_TYPE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_SEGMENT_START">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_SEGMENT_START</h4>
<pre>public static final&nbsp;java.lang.String P_SEGMENT_START</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_SEGMENT_START">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_SEGMENT_END">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_SEGMENT_END</h4>
<pre>public static final&nbsp;java.lang.String P_SEGMENT_END</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_SEGMENT_END">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_SEGMENT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_SEGMENT</h4>
<pre>public static final&nbsp;java.lang.String P_SEGMENT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_SEGMENT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="P_DUPLICATE_RETRIES">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_DUPLICATE_RETRIES</h4>
<pre>public static final&nbsp;java.lang.String P_DUPLICATE_RETRIES</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.P_DUPLICATE_RETRIES">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_ONE_POINT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_ONE_POINT</h4>
<pre>public static final&nbsp;int C_ONE_POINT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_ONE_POINT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_ONE_POINT_NO_NOP">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_ONE_POINT_NO_NOP</h4>
<pre>public static final&nbsp;int C_ONE_POINT_NO_NOP</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_ONE_POINT_NO_NOP">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_TWO_POINT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_TWO_POINT</h4>
<pre>public static final&nbsp;int C_TWO_POINT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_TWO_POINT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_TWO_POINT_NO_NOP">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_TWO_POINT_NO_NOP</h4>
<pre>public static final&nbsp;int C_TWO_POINT_NO_NOP</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_TWO_POINT_NO_NOP">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_ANY_POINT">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_ANY_POINT</h4>
<pre>public static final&nbsp;int C_ANY_POINT</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_ANY_POINT">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_LINE_RECOMB">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_LINE_RECOMB</h4>
<pre>public static final&nbsp;int C_LINE_RECOMB</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_LINE_RECOMB">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_INTERMED_RECOMB">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_INTERMED_RECOMB</h4>
<pre>public static final&nbsp;int C_INTERMED_RECOMB</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_INTERMED_RECOMB">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_SIMULATED_BINARY">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_SIMULATED_BINARY</h4>
<pre>public static final&nbsp;int C_SIMULATED_BINARY</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_SIMULATED_BINARY">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_NONE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_NONE</h4>
<pre>public static final&nbsp;int C_NONE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_NONE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_GEOMETRIC">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_GEOMETRIC</h4>
<pre>public static final&nbsp;int C_GEOMETRIC</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_GEOMETRIC">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="C_UNIFORM">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>C_UNIFORM</h4>
<pre>public static final&nbsp;int C_UNIFORM</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.VectorSpecies.C_UNIFORM">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="duplicateRetries">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>duplicateRetries</h4>
<pre>protected&nbsp;int[] duplicateRetries</pre>
<div class="block">How often do we retry until we get a non-duplicate gene?</div>
</li>
</ul>
<a name="mutationProbability">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>mutationProbability</h4>
<pre>protected&nbsp;double[] mutationProbability</pre>
<div class="block">Probability that a gene will mutate, per gene.
        This array is one longer than the standard genome length.
        The top element in the array represents the parameters for genes in
        genomes which have extended beyond the genome length.</div>
</li>
</ul>
<a name="crossoverProbability">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>crossoverProbability</h4>
<pre>public&nbsp;double crossoverProbability</pre>
<div class="block">Probability that a gene will cross over -- ONLY used in V_ANY_POINT crossover</div>
</li>
</ul>
<a name="crossoverType">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>crossoverType</h4>
<pre>public&nbsp;int crossoverType</pre>
<div class="block">What kind of crossover do we have?</div>
</li>
</ul>
<a name="genomeSize">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genomeSize</h4>
<pre>public&nbsp;int genomeSize</pre>
<div class="block">How big of a genome should we create on initialization?</div>
</li>
</ul>
<a name="crossoverDistributionIndex">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>crossoverDistributionIndex</h4>
<pre>public&nbsp;int crossoverDistributionIndex</pre>
<div class="block">What should the SBX distribution index be?</div>
</li>
</ul>
<a name="genomeResizeAlgorithm">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genomeResizeAlgorithm</h4>
<pre>public&nbsp;int genomeResizeAlgorithm</pre>
<div class="block">How should we reset the genome?</div>
</li>
</ul>
<a name="minInitialSize">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>minInitialSize</h4>
<pre>public&nbsp;int minInitialSize</pre>
<div class="block">What's the smallest legal genome?</div>
</li>
</ul>
<a name="maxInitialSize">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>maxInitialSize</h4>
<pre>public&nbsp;int maxInitialSize</pre>
<div class="block">What's the largest legal genome?</div>
</li>
</ul>
<a name="genomeIncreaseProbability">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genomeIncreaseProbability</h4>
<pre>public&nbsp;double genomeIncreaseProbability</pre>
<div class="block">With what probability would our genome be at least 1 larger than it is now during initialization?</div>
</li>
</ul>
<a name="chunksize">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>chunksize</h4>
<pre>public&nbsp;int chunksize</pre>
<div class="block">How big of chunks should we define for crossover?</div>
</li>
</ul>
<a name="lineDistance">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>lineDistance</h4>
<pre>public&nbsp;double lineDistance</pre>
<div class="block">How far along the long a child can be located for line or intermediate recombination</div>
</li>
</ul>
<a name="dynamicInitialSize">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>dynamicInitialSize</h4>
<pre>public&nbsp;boolean dynamicInitialSize</pre>
<div class="block">Was the initial size determined dynamically?</div>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="VectorSpecies()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>VectorSpecies</h4>
<pre>public&nbsp;VectorSpecies()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="mutationProbability(int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>mutationProbability</h4>
<pre>public&nbsp;double&nbsp;mutationProbability(int&nbsp;gene)</pre>
</li>
</ul>
<a name="duplicateRetries(int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>duplicateRetries</h4>
<pre>public&nbsp;int&nbsp;duplicateRetries(int&nbsp;gene)</pre>
</li>
</ul>
<a name="defaultBase()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultBase</h4>
<pre>public&nbsp;<a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;defaultBase()</pre>
<div class="block"><strong>Description copied from interface:&nbsp;<code><a href="../../ec/Prototype.html#defaultBase()">Prototype</a></code></strong></div>
<div class="block">Returns the default base for this prototype.
        This should generally be implemented by building off of the static base()
        method on the DefaultsForm object for the prototype's package. This should
        be callable during setup(...).</div>
</li>
</ul>
<a name="setupGenome(ec.EvolutionState, ec.util.Parameter)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setupGenome</h4>
<pre>protected&nbsp;void&nbsp;setupGenome(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
               <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</pre>
</li>
</ul>
<a name="setup(ec.EvolutionState, ec.util.Parameter)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setup</h4>
<pre>public&nbsp;void&nbsp;setup(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
         <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Species.html#setup(ec.EvolutionState, ec.util.Parameter)">Species</a></code></strong></div>
<div class="block">The default version of setup(...) loads requested pipelines and calls setup(...) on them and normalizes their probabilities.  
        If your individual prototype might need to know special things about the species (like parameters stored in it),
        then when you override this setup method, you'll need to set those parameters BEFORE you call super.setup(...),
        because the setup(...) code in Species sets up the prototype.</div>
<dl>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Prototype.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../ec/Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Setup.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../ec/Setup.html" title="interface in ec">Setup</a></code></dd>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Species.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Species.html" title="class in ec">Species</a></code></dd>
<dt><span class="strong">See Also:</span></dt><dd><a href="../../ec/Prototype.html#setup(ec.EvolutionState, ec.util.Parameter)"><code>Prototype.setup(EvolutionState,Parameter)</code></a></dd></dl>
</li>
</ul>
<a name="loadParametersForGene(ec.EvolutionState, int, ec.util.Parameter, ec.util.Parameter, java.lang.String)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>loadParametersForGene</h4>
<pre>protected&nbsp;void&nbsp;loadParametersForGene(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                         int&nbsp;index,
                         <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base,
                         <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;def,
                         java.lang.String&nbsp;postfix)</pre>
<div class="block">Called when VectorSpecies is setting up per-gene and per-segment parameters.  The index
        is the current gene whose parameter is getting set up.  The Parameters in question are the
        bases for the gene.  The postfix should be appended to the end of any parameter looked up
        (it often contains a number indicating the gene in question), such as
        state.parameters.exists(base.push(P_PARAM).push(postfix), def.push(P_PARAM).push(postfix)

        <p>If you override this method, be sure to call super(...) at some point, ideally first.</div>
</li>
</ul>
<a name="initializeGenomeSegmentsByStartIndices(ec.EvolutionState, ec.util.Parameter, ec.util.Parameter, int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initializeGenomeSegmentsByStartIndices</h4>
<pre>protected&nbsp;void&nbsp;initializeGenomeSegmentsByStartIndices(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                                          <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base,
                                          <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;def,
                                          int&nbsp;numSegments)</pre>
<div class="block">Looks up genome segments using start indices.  Segments run up to the next declared start index.</div>
</li>
</ul>
<a name="initializeGenomeSegmentsByEndIndices(ec.EvolutionState, ec.util.Parameter, ec.util.Parameter, int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initializeGenomeSegmentsByEndIndices</h4>
<pre>protected&nbsp;void&nbsp;initializeGenomeSegmentsByEndIndices(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                                        <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base,
                                        <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;def,
                                        int&nbsp;numSegments)</pre>
<div class="block">Looks up genome segments using end indices.  Segments run from the previously declared end index.</div>
</li>
</ul>
<a name="newIndividual(ec.EvolutionState, int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>newIndividual</h4>
<pre>public&nbsp;<a href="../../ec/Individual.html" title="class in ec">Individual</a>&nbsp;newIndividual(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                       int&nbsp;thread)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Species.html#newIndividual(ec.EvolutionState, int)">Species</a></code></strong></div>
<div class="block">Provides a brand-new individual to fill in a population.  The default form
        simply calls clone(), creates a fitness, sets evaluated to false, and sets
        the species.  If you need to make a more custom genotype (as is the case
        for GPSpecies, which requires a light rather than deep clone), 
        you will need to override this method as you see fit.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Species.html#newIndividual(ec.EvolutionState, int)">newIndividual</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Species.html" title="class in ec">Species</a></code></dd>
</dl>
</li>
</ul>
<a name="fill(long[], long)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>fill</h4>
<pre>protected&nbsp;long[]&nbsp;fill(long[]&nbsp;array,
          long&nbsp;val)</pre>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</li>
</ul>
<a name="fill(int[], int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>fill</h4>
<pre>protected&nbsp;int[]&nbsp;fill(int[]&nbsp;array,
         int&nbsp;val)</pre>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</li>
</ul>
<a name="fill(boolean[], boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>fill</h4>
<pre>protected&nbsp;boolean[]&nbsp;fill(boolean[]&nbsp;array,
             boolean&nbsp;val)</pre>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</li>
</ul>
<a name="fill(double[], double)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>fill</h4>
<pre>protected&nbsp;double[]&nbsp;fill(double[]&nbsp;array,
            double&nbsp;val)</pre>
<div class="block">Utility method: fills the array with the given value and returns it.</div>
</li>
</ul>
<a name="contains(boolean[], boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>contains</h4>
<pre>protected&nbsp;int&nbsp;contains(boolean[]&nbsp;array,
           boolean&nbsp;val)</pre>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</li>
</ul>
<a name="contains(long[], long)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>contains</h4>
<pre>protected&nbsp;int&nbsp;contains(long[]&nbsp;array,
           long&nbsp;val)</pre>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</li>
</ul>
<a name="contains(int[], int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>contains</h4>
<pre>protected&nbsp;int&nbsp;contains(int[]&nbsp;array,
           int&nbsp;val)</pre>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</li>
</ul>
<a name="contains(double[], double)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>contains</h4>
<pre>protected&nbsp;int&nbsp;contains(double[]&nbsp;array,
           double&nbsp;val)</pre>
<div class="block">Utility method: returns the first array slot which contains the given value, else -1.</div>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-all.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector"><span class="strong">Prev Class</span></a></li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../../index.html?ec/vector/VectorSpecies.html" target="_top">Frames</a></li>
<li><a href="VectorSpecies.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field_summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field_detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
